src/lib/viewers/controls/page/PageControlsForm.tsx

```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import { decodeKeydown } from '../../../util';
+import './PageControlsForm.scss';
+
+export type Props = {
+    onPageSubmit: (newPageNumber: number) => void;
+    pageCount: number;
+    pageNumber: number;
+};
+
+export const ENTER = 'Enter';
+export const ESCAPE = 'Escape';
+
+export default function PageControlsForm({ onPageSubmit, pageNumber, pageCount }: Props): JSX.Element {
+    const [isInputShown, setIsInputShown] = React.useState(false);
+    const [inputValue, setInputValue] = React.useState(pageNumber);
+    const isButtonFocus = React.useRef(false);
+    const buttonElRef = React.useRef<HTMLButtonElement>(null);
+    const inputElRef = React.useRef<HTMLInputElement>(null);
+
+    const setPage = (allowRetry = false): void => {
+        if (!Number.isNaN(inputValue) && inputValue >= 1 && inputValue <= pageCount && inputValue !== pageNumber) {
+            onPageSubmit(inputValue);
+        } else {
+            setInputValue(pageNumber); // Reset the invalid input value to the current page
+
+            if (allowRetry) {
+                isButtonFocus.current = true;
+            }
+        }
+
+        setIsInputShown(false);
+    };
+
+    const handleNumInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+        setInputValue(parseInt(event.target.value, 10));
+    };
+
+    const handleNumInputBlur = (): void => {
+        setPage();
+    };
+
+    const handleNumInputKeyDown = (event: React.KeyboardEvent<HTMLInputElement>): void => {
+        const key = decodeKeydown(event);
+
+        switch (key) {
+            case ENTER:
+                event.stopPropagation();
+                event.preventDefault();
+
+                setPage(true);
+                break;
+            case ESCAPE:
+                event.stopPropagation();
+                event.preventDefault();
+
+                setIsInputShown(false);
+                isButtonFocus.current = true;
```

Comment:
It may not make a difference, but we should do this before we call `setIsInputShown`.